tooling/ipns/record.go

```diff
@@ -41,10 +51,46 @@ func (i *IpnsRecord) Validity() time.Time {
 }
 
 func (i *IpnsRecord) Valid() error {
-	id, err := peer.Decode(i.key)
+	return ipns.ValidateWithPeerID(i.id, i.pb)
+}
+
+func (i *IpnsRecord) idV1(codec multicodec.Code, base mbase.Encoding) (string, error) {
+	c := peer.ToCid(i.id)
+	c = cid.NewCidV1(uint64(codec), c.Hash())
+	s, err := c.StringOfBase(base)
+	if err != nil {
+		return "", err
+	}
+	return s, nil
+}
+
+func (i *IpnsRecord) IntoCID(codec multicodec.Code, base mbase.Encoding) string {
+	s, err := i.idV1(codec, base)
 	if err != nil {
-		return err
+		panic(err)
+	}
+	return s
+}
+
+func (i *IpnsRecord) IdV0() string {
+	if strings.HasPrefix(i.key, "Qm") || strings.HasPrefix(i.key, "1") {
+		return i.key
 	}
+	panic("not a v0 id")
+	// This code won't work if key starts with "1", this is an identity multihash, which
+	// makes `NewCidV0` panic.
+	// c := peer.ToCid(i.id)
+	// s, err := cid.NewCidV0(c.Hash()).StringOfBase(mbase.Base58BTC)
+	// if err != nil {
+	// 	panic(err)
+	// }
+	// return s
+}
```

Comment:
Highlight: I couldn't find a reusable way to do this, so I copied some code out of kubo.